src/core/server/plugins/plugin.ts

```diff
@@ -168,8 +169,8 @@ export type PluginInitializer<
 export class PluginWrapper<
   TSetup = unknown,
   TStart = unknown,
-  TPluginsSetup extends Record<PluginName, unknown> = Record<PluginName, unknown>,
-  TPluginsStart extends Record<PluginName, unknown> = Record<PluginName, unknown>
+  TPluginsSetup extends EmptyPluginContracts = EmptyPluginContracts,
```

Comment:
I just realised that `extends EmptyPluginContracts` refers to an empty interface, not `object`. with this change we have a problem that any real type like `number`, `string`, etc. satisfy this conditions. now I think we should rollback to `Record<string, unknown>`

https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgKIFsAOYCeAFAGwFcBzUAYQHtwpEwBnZAbwF8AoN3TFAQQB4AKgD5kAXmQDkEAB6QQAE0YZs+YmRBUadRgH5kAchiVK+5AC4DAIzhR9Abg5cUAITHJ+9MFFAkhd5AD0AchGlJw43MjkbvwgROiW0H6BwaHhkQAiMXysycj5KSHG6Sio2QBulMDyeYXWUI4RvPzCbpIycorIAEoQCJRQ8nye3iAkADTIRCAA1iCUAO4gInqGxqYW+vX2jZHOruI8Hl4+yUHI9SVR0Yex8YlQZ8GXTsgZWYc5LLUF52mvqDKn0q1R+53qQA

Reply:
@restrry 

I'm not sure if that is something we should worry about, but if you want to fix that you can do:

```ts
type EmptyPluginContracts = {} & object;
```

or use a branded type:

```ts
type EmptyPluginContracts = {} & { __brandPluginContract?: 'yup' };
```

or use a branded type from some lib like [`utility-types`](https://github.com/piotrwitek/utility-types#brandt-u):

```ts
type EmptyPluginContracts = Brand<{}, "PluginContracts">
```

Reply:
> I'm not sure if that is something we should worry about

yes, right now it doesn't specify a type of a contract.
```ts
TPluginsSetup extends object = {},
```
should be enough

Reply:
That does not seem to work.

![image](https://user-images.githubusercontent.com/9773803/60510654-3a7e5800-9cd0-11e9-8863-0b3e4a1b346b.png)

Reply:
hm. it falls back to the empty interface as well. either options should work
```typescript
TPluginsSetup extends object = object,
TPluginsSetup extends Record<string, unknown> = Record<string, unknown>,
```

Reply:
@restrry 

After playing with it for a bit, I think I have [something that cover "all" 4 use cases](https://www.typescriptlang.org/play/#code/C4TwDgpgBACgTgSwLYOAgbtAvAKAJAA+UAdgK4A25+RpxAJhAGYLER3VQDOwixA5hzJIARhDgdhAe0nkIAQ2IdOIETIkI+LYAG4cOUJCgA5ScHjJUGCAB4AKgD4oWKLagQAHsAj1OsRCjRMKAB+EghMOCgALhddHC0xRjkAY2gYclI+KDsAZQhgUjA3T286XwAKdEkEOigiEzN-S0xrAG8AX3sASicoSWEAKwhk4EdWnChJrnzC4PKumNs8gqKPLx8-C0DoUNYI6Jdlwt12vQB6M6gAbQBGADoAXSgAVWIAa2JJAHdiKDA4SSQOBoCC+ZIKT7AKCiKApVKcThsO7xYheOBJVJQI5gG5QcZTKCMaQxISiOAnHDJchyBFQACyIHSmRYuOQYFkSG8wF8TL41mxNzGEymiJW8zxwoJkzgMzgv1aRMkMRuACYAMztXQE06nVhfemMjKaYg3eZ3UWFM2K7RQC7ZAC0jq4pGS8M4OD1Bt5LLNFrAZuEcnJtsu1kd9rccAB4hwdquKsesCNLCgyUkqLgKSh4OIkOh0Dkfya2ygACJSWJS1ADBBkTWsTMwCrehXyZTqbSGd7iM22RyuTyjfzGyqhQS-eLWiGHU6xNHJVKZQU5VBVWqtVMdedLlc1YmAJKMP7J35pjNZqAIXx5ziQZIIZhsAA0l6h3AQlFjlxzeZhhf+WxWGWrZVjWyJUjSvhdieaqXkg7IQJyqKDpkErjo2k4LgSS6kCua4bpMW5ftcAAsibdqmChQAAFnIQSfKm6Y8Fm4EdlBhrMsQJFwQhSHckmmTWFUNRjiKGE9OMC5EXoOBAA):

1. Unknown properties cannot be accessed.
1. Plugin contract cannot be a primitive "number" type.
1. If plugin contract is not specified, it still cannot be a primitive like "number".
1. Plugin can have no contract.

![image](https://user-images.githubusercontent.com/9773803/60515102-3277e580-9cdb-11e9-9463-4fad7526526f.png)

Reply:
Ah, actually `object/object` this way seems to cover those cases too:

```ts
TPluginsSetup extends (object | void) = object
```

If you do that same thing with `Record` it does not work.

```ts
TPluginsSetup extends (Record<string, unknown> | void) = Record<string, unknown>
```

OK, I know why. Because `type NonPrimitive = object;` 😄

Reply:
@restrry so, I'm gonna implement

```ts
TPluginsSetup extends (object | void) = object
```

if you are okay with that?

Reply:
should be ok